pkg_resources/__init__.py

```diff
@@ -65,40 +66,44 @@
 from os import open as os_open
 from os.path import isdir, split
 
-from pkg_resources.extern.jaraco.text import (
-    yield_lines,
-    drop_comment,
-    join_continuation,
-)
+if not TYPE_CHECKING:
+    from pkg_resources.extern.jaraco.text import (
+        yield_lines,
+        drop_comment,
+        join_continuation,
+    )
 
-from pkg_resources.extern import platformdirs
-from pkg_resources.extern import packaging
-
-__import__('pkg_resources.extern.packaging.version')
-__import__('pkg_resources.extern.packaging.specifiers')
-__import__('pkg_resources.extern.packaging.requirements')
-__import__('pkg_resources.extern.packaging.markers')
-__import__('pkg_resources.extern.packaging.utils')
-
-# declare some globals that will be defined later to
-# satisfy the linters.
-require = None
-working_set = None
-add_activation_listener = None
-resources_stream = None
-cleanup_resources = None
-resource_dir = None
-resource_stream = None
-set_extraction_path = None
-resource_isdir = None
-resource_string = None
-iter_entry_points = None
-resource_listdir = None
-resource_filename = None
-resource_exists = None
-_distribution_finders = None
-_namespace_handlers = None
-_namespace_packages = None
+    from pkg_resources.extern import platformdirs
+    from pkg_resources.extern import packaging
+
+    __import__('pkg_resources.extern.packaging.version')
+    __import__('pkg_resources.extern.packaging.specifiers')
+    __import__('pkg_resources.extern.packaging.requirements')
+    __import__('pkg_resources.extern.packaging.markers')
+    __import__('pkg_resources.extern.packaging.utils')
+else:
+    # Replicates the imports above in a way that can be understood by type-checkers
+    from jaraco.text import (
+        yield_lines,
+        drop_comment,
+        join_continuation,
+    )
+    import platformdirs
+    import packaging
+    import packaging.version
+    import packaging.specifiers
+    import packaging.requirements
+    import packaging.markers
+    import packaging.utils
```

Comment:
FWIW, I'd rather loose on type safety, and loose on implicit return types (until all public API return types are explicit), than going for a mypy plugin. Which also only works on mypy.

I'll probably end up keeping those changes in my local working branches since it helps me validate the types and understand intended behaviour. Even if it doesn't end up in `main`.

Reply:
Hang on, now that there's a `cog` script/template thingy, could I extend it to do https://github.com/pypa/setuptools/pull/3979#discussion_r1367960497 again but automated this time?

Reply:
Yeah, I don't see why not. But currently there is a problem with the script that I am fixing in https://github.com/pypa/setuptools/pull/4353, so it might make sense to wait for it.

Reply:
Anyway, I will defer the final decision about any workarounds for the lack of support of `MetaPathFinder`s in type checkers to Jason.

Reply:
Yeah I'm likely gonna do this last anyway. It's just good to know I have more options to try and suggest.